File: documentation-examples-java/src/java/dev/typr/foundations/docs/core/FragmentBuilding.java
package dev.typr.foundations.docs.core;

import dev.typr.foundations.Connection;
import dev.typr.foundations.Fragment;
import dev.typr.foundations.PgTypes;
import dev.typr.foundations.RowCodec;
import java.time.Instant;
import java.util.List;

@SuppressWarnings("unused")
public class FragmentBuilding {
  record User(Integer id, String name, String status, Instant createdAt) {}

  RowCodec<User> userCodec =
      RowCodec.<User>builder()
          .field(PgTypes.int4, User::id)
          .field(PgTypes.text, User::name)
          .field(PgTypes.text, User::status)
          .field(PgTypes.timestamptz, User::createdAt)
          .build(User::new);

  Connection connection = null; // placeholder
  Integer userId = 1;
  Instant cutoffDate = Instant.now();

  // start
  Fragment query =
      Fragment.of(
              """
              SELECT * FROM users
              WHERE id =
              """)
          .value(PgTypes.int4, userId)
          .append(" AND status = ")
          .value(PgTypes.text, "active")
          .append(" AND created_at > ")
          .value(PgTypes.timestamptz, cutoffDate);

  // Execute safely — parameters are bound, not interpolated
  List<User> users = query.query(userCodec.all()).run(connection);
  // stop
}

File: documentation-examples-scala/src/scala/dev/typr/foundationssc/docs/core/FragmentBuilding.scala
package dev.typr.foundationssc.docs.core
import dev.typr.foundationssc.*
import dev.typr.foundationssc.Fragment.*
import dev.typr.foundationssc.data.*

import java.time.Instant

@SuppressWarnings(Array("unused"))
object FragmentBuilding:
  case class User(id: Int, name: String, status: String, createdAt: Instant)

  val userCodec: RowCodec[User] = RowCodec
    .builder[User]()
    .field(PgTypes.int4)(_.id)
    .field(PgTypes.text)(_.name)
    .field(PgTypes.text)(_.status)
    .field(PgTypes.timestamptz)(_.createdAt)
    .build(User.apply)

  val userId: Integer = 1
  val cutoffDate: Instant = Instant.now()

  // start
  val query: Fragment =
    sql"""SELECT * FROM users
          WHERE id = ${PgTypes.int4(userId)}
          AND status = ${PgTypes.text("active")}
          AND created_at > ${PgTypes.timestamptz(cutoffDate)}"""

  // Parameters are bound, not interpolated
  def users(using Connection): List[User] =
    query.query(userCodec.all()).run
  // stop
